### PR TITLE
feat(trie): stored nibbles

### PR DESCRIPTION
**Pull request by rkrasiuk**
Introduces `StoredNibbles` and `StoredNibblesSubKey`, structs which will be used as key and subkey in trie tables.
Moves `Nibbles` implementation to the `trie` crate.

### Summary
🗜️🔀🧩

Refactor trie storage and encoding logic using new types `StoredNibbles` and `StoredNibblesSubKey` that implement `Encode` and `Decode` traits. Move the `Nibbles` type from the `reth_primitives` crate to the `reth_trie` crate to avoid circular dependencies and improve trie operations. Update imports and re-exports accordingly. Add `derive_more` crate dependency for convenience.

> _We're breaking the chains of the primitives_
> _We're forging the nibbles of the trie_
> _We're compressing the keys with new methods_
> _We're refactoring the code to the sky_

### Walkthrough
*  Refactor the trie storage and encoding logic by introducing new types `StoredNibbles` and `StoredNibblesSubKey` to represent the nibbles of the trie keys and subkeys in a compact and efficient way ([link](https://github.com/paradigmxyz/reth/pull/2182/files?diff=unified&w=0#diff-2b20291b814762d5362752c8e8f661c3a5f84f6ad9bf7f3be5bbdaaa9c38e95aL4-R4), [link](https://github.com/paradigmxyz/reth/pull/2182/files?diff=unified&w=0#diff-71a31b8cbc74276c0eff6604c5bf17d52e13a6b35ff8c6af9c771a587a10f1b4L7-R10), [link](https://github.com/paradigmxyz/reth/pull/2182/files?diff=unified&w=0#diff-71a31b8cbc74276c0eff6604c5bf17d52e13a6b35ff8c6af9c771a587a10f1b4R46-R47), [link](https://github.com/paradigmxyz/reth/pull/2182/files?diff=unified&w=0#diff-b485e429770783ad8654467279496e8e13c0d307213b4a5682c80ecb47287ef9R2-R10), [link](https://github.com/paradigmxyz/reth/pull/2182/files?diff=unified&w=0#diff-b485e429770783ad8654467279496e8e13c0d307213b4a5682c80ecb47287ef9R101-R136)).

### CODE REVIEW

File: crates/primitives/src/trie/nibbles.rs [hunk 1]
```diff
@@ -1,290 +1,48 @@
...
-            // Check first byte
-            assert_eq!(extension_flag & 0x20, 0);
-            assert_eq!(input_is_odd, (extension_flag & 0x10) != 0);
-            if input_is_odd {
-                assert_eq!(extension_flag & 0x0f, *input.first().unwrap());
-            }
-        }
+impl Compact for StoredNibblesSubKey {
+    fn to_compact(self, buf: &mut impl bytes::BufMut) -> usize {
+        assert!(self.inner.len() <= 32);
```
Comment (joshieDo) [inline crates/primitives/src/trie/nibbles.rs:32]: necessary?

File: crates/primitives/src/trie/nibbles.rs [hunk 2]
```diff
@@ -1,290 +1,48 @@
...
+        let mut padded = vec![0; 32];
+        padded[..self.inner.len()].copy_from_slice(&self.inner[..]);
+        buf.put_slice(&padded);
```
Comment (joshieDo) [inline crates/primitives/src/trie/nibbles.rs:35]: i guess it's not much of a difference, but can we copy `self.inner` to `buf`  directly, and just copy the necessary padded 0 bytes from a fixed array.

File: crates/storage/db/src/tables/models/mod.rs [hunk 3]
```diff
@@ -95,3 +98,39 @@ impl Decode for String {
         String::from_utf8(value.as_ref().to_vec()).map_err(|_| Error::DecodeError)
     }
 }
+
+impl Encode for StoredNibbles {
+    type Encoded = Vec<u8>;
+
+    // Delegate to the Compact implementation
+    fn encode(self) -> Self::Encoded {
+        let mut buf = Vec::new();
```
Comment (joshieDo) [inline crates/storage/db/src/tables/models/mod.rs:107]: Vec::with_capacity ?

File: crates/storage/db/src/tables/models/mod.rs [hunk 4]
```diff
@@ -95,3 +98,39 @@ impl Decode for String {
...
+        Ok(Self::from_compact(buf, buf.len()).0)
+    }
+}
+
+impl Encode for StoredNibblesSubKey {
+    type Encoded = Vec<u8>;
+
+    // Delegate to the Compact implementation
+    fn encode(self) -> Self::Encoded {
+        let mut buf = Vec::new();
```
Comment (joshieDo) [inline crates/storage/db/src/tables/models/mod.rs:125]: Vec::with_capacity(33) ?

File: crates/primitives/src/trie/nibbles.rs [hunk 5]
```diff
@@ -1,290 +1,48 @@
...
+        buf.put_u8(self.inner.len() as u8);
+        33 // 32 + 1
```
Comment (mattsse) [inline crates/primitives/src/trie/nibbles.rs:37]:
```suggestion
        32 + 1
```

File: crates/primitives/src/trie/nibbles.rs (hunk 1 above)
Comment (rkrasiuk) [reply crates/primitives/src/trie/nibbles.rs]: yes, but thanks for drawing my attention, should be 64

File: crates/trie/src/nibbles.rs [hunk 6]
```diff
@@ -0,0 +1,296 @@
...
+    Ord,
+    Index,
+    From,
+    Deref,
+    DerefMut,
+)]
+pub struct Nibbles {
+    /// The inner representation of the nibble sequence.
+    pub hex_data: Vec<u8>,
+}
```
Comment (gakonst) [inline crates/trie/src/nibbles.rs:31]: We'll want some From/Into conversions for Stored nibbles, but looks good

**Review by gakonst — APPROVED** [review]

**Pull request merged**